README
tomcast
=======
tomcast reads mpeg transport streams over http or udp (multicast or
unicast) and outputs them to chosen multicast group. Basically a simple
http2multicast daemon designed to work 24/7.

License
=======
tsdecrypt is released under GNU GPL v2.

Development
===========
The development is done using git. tomcast repository is hosted
at http://github.com/gfto/tomcast

To clone the repository issue the following commands:

   git clone git://github.com/gfto/tomcast.git
   cd tomcast
   git submodule init
   git submodule update
   make

The code is developed and tested under modern Linux. It is also
compiled from time to time under OS X but is not tested there.

To see all Makefile targets run 'make help'.

Updating the code
=================
To update cloned tomcast, go to the directory where the repository
is cloned and run the following commands:

   git fetch origin
   git merge origin/master
   git submodule update
   make clean all

tomcast's master branch should always be useful so it is safe to
use it instead of official release. The master branch will always
be better than any released version.

Releases
========
Official releases can be downloaded from tomcast home page which is

   http://georgi.unixsol.org/programs/tomcast/

Contact
=======
For patches, bug reports, complaints and so on send e-mail to

   Georgi Chorbadzhiyski <georgi@unixsol.org>


ChangeLog
vNEXT | xx xxx xxxx
  * Add support for handling of 301/302 HTTP redirects
  * Add support for detecting encrypted input and switching sources
    if the input is encrypted (use -E option).

v1.30 | 21 Dec 2016
  * Add web access for monitoring and reconfiguration

v1.20 | 17 Aug 2013
  * First publicly released version.

v1.15 | 13 Sep 2012
  * Add option to send reset packets when source have been reconnected.
    These packets prevent VLC from hanging.

v1.14 | 16 May 2012
  * Set send buffer size.

v1.13 | 01 Mar 2012
  * Use mpeg_sync on connect.
  * Update libfuncs to latest version from upstream repo.

v1.12 | 10 Dec 2010
  * Allow URL to contain anything

v1.11 | 21 Jul 2010
  * Send X-Smart-Client: yes in GET request

v1.10 | 24 Feb 2010
  * Check for read errors and reconnect
  * Check for write errors and reconnect udp socket if there are any

v1.00 | 09 Feb 2010
  * First version.


config.h
/*
 * mptsd configuration header file
 * Copyright (C) 2010-2011 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */
#ifndef CONFIG_H
#define CONFIG_H

#include <pthread.h>
#include <arpa/inet.h>
#include <netinet/in.h>

#include "libfuncs/libfuncs.h"

typedef enum { udp_sock, tcp_sock } channel_source;

typedef struct {
	channel_source sproto;
	char *proto;
	char *host;
	char *path;
	unsigned int port;
} CHANSRC;

#define MAX_CHANNEL_SOURCES 8

typedef struct {
	char *name;
	char *source; /* Full source url */
	char *sources[MAX_CHANNEL_SOURCES];
	uint8_t num_src;
	uint8_t curr_src;
	char *dest_host;
	int dest_port;
} CHANNEL;

typedef struct {
	char  *name;
	CHANNEL *channel;
	int sock;				/* Server socket */
	struct sockaddr_in src_sockname;
	int clientsock;			/* The udp socket */
	struct sockaddr_in dst_sockname;
	int reconnect:1,		/* Set to 1 to force proxy reconnect */
	    connected:1,		/* It's set to 1 when proxy is connected and serving clients */
	    dienow:1,			/* Stop serving clients and exit now */
	    freechannel:1;		/* Free channel data on object free (this is used in chanconf) */
	int cookie;				/* Used in chanconf to determine if the restreamer is alrady checked */
	pthread_t thread;
	pthread_rwlock_t lock;
	time_t conn_ts;
	uint64_t read_bytes;
	int64_t last_decrypted_input_ts;
	char status[64];
} RESTREAMER;


struct config {
	char				*ident;
	char				*pidfile;

	int					syslog_active;
	char				*logident;
	char				*loghost;
	int					logport;

	struct sockaddr_in	server;
	char				*server_addr;
	int					server_port;
	int					server_socket;
	pthread_t			server_thread;
	int					allow_encrypted_input;

	char				*channels_file;

	LIST				*chanconf;
	LIST				*restreamer;

	pthread_mutex_t		channels_lock;
};

extern void do_reconnect(int sig);
extern void do_reconf(int sig);
extern struct config *get_config(void);

#endif


web_pages.c
/*
 * mptsd internal web pages
 * Copyright (C) 2010-2011 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <ctype.h>
#include <unistd.h>

#include "libfuncs/io.h"
#include "libfuncs/log.h"
#include "libfuncs/list.h"
#include "libfuncs/http_response.h"

#include "config.h"

extern struct config *config;

void cmd_index(int clientsock) {
	send_200_ok(clientsock);
	send_header_textplain(clientsock);
	fdputs(clientsock, "\nHi from tomcast.\n");
}

void cmd_status(int clientsock) {
	send_200_ok(clientsock);
	send_header_textplain(clientsock);
	fdputs(clientsock, "\n");

	LNODE *l, *tmp;
	struct config *cfg = get_config();

	time_t now = time(NULL);
	fdputsf(clientsock, "%-10s %-20s %8s %10s %-18s %-64s %s\n",
		"# Status",
		"DestAddr",
		"ConnTime",
		"ReadBytes",
		"ChanName",
		"ChanSource",
		"ProxyStatus"
	);
	pthread_mutex_lock(&cfg->channels_lock);
	list_lock(cfg->restreamer);
	list_for_each(cfg->restreamer, l, tmp) {
		char dest[32];
		int conn_status = 0;
		RESTREAMER *r = l->data;
		pthread_rwlock_rdlock(&r->lock);
		snprintf(dest, sizeof(dest), "%s:%d", r->channel->dest_host, r->channel->dest_port);
		if (r->connected && r->conn_ts > 0 && r->read_bytes >= 1316)
			conn_status = 1;
		fdputsf(clientsock, "%-10s %-20s %8lu %10llu %-18s %-64s %s\n",
			conn_status ? "CONN_OK" : "CONN_ERROR",
			dest,
			r->conn_ts ? now - r->conn_ts : 0,
			r->read_bytes,
			r->channel->name,
			r->channel->source,
			r->status
		);
		pthread_rwlock_unlock(&r->lock);
	}
	list_unlock(cfg->restreamer);
	pthread_mutex_unlock(&cfg->channels_lock);
}

void cmd_getconfig(int clientsock) {
	send_200_ok(clientsock);
	send_header_textplain(clientsock);
	fdputs(clientsock, "\n");

	LNODE *l, *tmp;
	struct config *cfg = get_config();

	pthread_mutex_lock(&cfg->channels_lock);
	list_lock(cfg->restreamer);
	list_for_each(cfg->restreamer, l, tmp) {
		RESTREAMER *r = l->data;
		pthread_rwlock_rdlock(&r->lock);
		int i;
		for (i = 0; i < r->channel->num_src; i++) {
			fdputsf(clientsock, "%s\t%s:%d\t%s\n",
				r->channel->name,
				r->channel->dest_host,
				r->channel->dest_port,
				r->channel->sources[i]
			);
		}
		pthread_rwlock_unlock(&r->lock);
	}
	list_unlock(cfg->restreamer);
	pthread_mutex_unlock(&cfg->channels_lock);
}

void cmd_reconnect(int clientsock) {
	send_200_ok(clientsock);
	send_header_textplain(clientsock);
	struct config *cfg = get_config();
	pthread_mutex_lock(&cfg->channels_lock);
	fdputsf(clientsock, "\nReconnecting %d inputs.\n", cfg->chanconf->items);
	pthread_mutex_unlock(&cfg->channels_lock);
	do_reconnect(1);
}

void cmd_reload(int clientsock) {
	send_200_ok(clientsock);
	send_header_textplain(clientsock);
	fdputs(clientsock, "\nReloading config\n");
	do_reconf(1);
}


web_pages.h
/*
 * mptsd internal web pages header file
 * Copyright (C) 2010-2011 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */
#ifndef WEB_PAGES_H
#define WEB_PAGES_H

void cmd_index(int clientsock);
void cmd_status(int clientsock);
void cmd_getconfig(int clientsock);
void cmd_reconnect(int clientsock);
void cmd_reload(int clientsock);

#endif


web_server.h
/*
 * mptsd internal web server header file
 * Copyright (C) 2010-2011 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */
#ifndef WEB_SERVER_H
# define WEB_SERVER_H

#include "config.h"

void web_server_start(struct config *conf);
void web_server_stop(struct config *conf);

#endif


web_server.c
/*
 * mptsd internal web server
 * Copyright (C) 2010-2011 Unix Solutions Ltd.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */
#include <stdlib.h>
#include <regex.h>
#include <errno.h>
#include <string.h>
#include <signal.h>
#include <arpa/inet.h>
#include <netinet/in.h>

#include "libfuncs/libfuncs.h"

#include "web_pages.h"
#include "web_server.h"

typedef struct req_info {
	int clientsock;
	struct sockaddr_in client;
} request_info;

extern int keep_going;

#define NEXT_CLIENT { FREE(path); FREE(buf); pthread_exit(0); }
#define SHUTDOWN_CLIENT { FREE(path); FREE(buf); shutdown_fd(&clientsock); pthread_exit(0); }
#define BUF_SIZE 1024

void *process_web_request(void *);

void *web_server_thread(void *data) {
	struct config *conf = data;
	while (keep_going) {
		struct sockaddr_in client;
		unsigned int clientlen = sizeof(client);
		int clientsock;
		clientsock = accept(conf->server_socket, (struct sockaddr *) &client, &clientlen);
		if (clientsock < 0) {
			if (conf->server_socket > -1)	// The server_socket is closed on exit, so do not report errors
				LOGf("ERROR : Failed to accept client fd: %i err: %s\n", clientsock, strerror(errno));
			if (errno==EMFILE || errno==ENFILE) /* No more FDs */
				break;
		} else {
			request_info *req;
			pthread_t req_thread;
			req = malloc(sizeof(request_info));
			if (!req) {
				log_perror("Can't allocate request_info", errno);
				continue;
			}
			req->clientsock = clientsock;
			req->client = client;
			if (pthread_create(&req_thread, NULL, (void *)&process_web_request, (void *)req)) {
				log_perror("Error creating request processing thread.", errno);
				exit(1);
			}
			pthread_detach(req_thread);
		}
	}

	pthread_exit(0);
}

void web_server_start(struct config *conf) {
	if (conf->server_socket > -1)
		pthread_create(&conf->server_thread, NULL, &web_server_thread, conf);
}

void web_server_stop(struct config *conf) {
	if (conf->server_socket > -1) {
		shutdown_fd(&conf->server_socket);
		pthread_join(conf->server_thread, NULL);
	}
}

void *process_web_request(void *in_req) {
	request_info *req = (request_info *)in_req;
	int clientsock = req->clientsock;
	regmatch_t res[3];
	char *path=NULL, *buf=NULL;
	FREE(req);

	signal(SIGPIPE, SIG_IGN);

	if (!keep_going)
		pthread_exit(0);

	buf = malloc(BUF_SIZE);
	if (!buf) {
		log_perror("Can't allocate buffer", errno);
		SHUTDOWN_CLIENT;
	}

	if (fdgetline(clientsock,buf,BUF_SIZE)<=0) {
		SHUTDOWN_CLIENT;
	}

	regex_t request_get;
	regcomp(&request_get, "^GET /([^ ]*) HTTP/1.*$", REG_EXTENDED);
	if (regexec(&request_get,buf,2,res,0)==REG_NOMATCH) {
		send_501_not_implemented(clientsock);
		SHUTDOWN_CLIENT;
	}

	buf[res[1].rm_eo]=0;
	chomp(buf+res[1].rm_so);
	if (buf[res[1].rm_eo-1]=='/') buf[res[1].rm_eo-1]=0;
	path = strdup(buf+res[1].rm_so);
	regfree(&request_get);

	while (fdgetline(clientsock,buf,BUF_SIZE) > 0) {
		if (buf[0] == '\n' || buf[0] == '\r') // End of headers
			break;
	}

	if (strlen(path) == 0) {
		cmd_index(clientsock);
	} else if (strstr(path,"getconfig")==path) {
		cmd_getconfig(clientsock);
	} else if (strstr(path,"reconnect")==path) {
		cmd_reconnect(clientsock);
	} else if (strstr(path,"reload")==path) {
		cmd_reload(clientsock);
	} else if (strstr(path,"status")==path) {
		cmd_status(clientsock);
	} else {
		send_404_not_found(clientsock);
	}

	SHUTDOWN_CLIENT;
}

